src/casync-http.c

```diff
@@ -11,67 +12,1066 @@
+                return 0;
+        }
+
+        if (!realloc_buffer_append(chunk_buffer, buffer, product)) {
+                log_oom();
+                return 0;
+        }
+
+        return product;
+}
+
+static void ca_chunk_downloader_free(CaChunkDownloader *dl) {
+        CURL *handle;
+        ChunkData *cd;
+
+        if (dl == NULL)
+                return;
+
+        while ((handle = queue_pop(dl->inprogress))) {
+                CURLMcode c;
+
+                c = curl_multi_remove_handle(dl->multi, handle);
+                if (c != CURLM_OK)
+                        log_error_curlm(c, "Failed to remove handle");
+
+                cd = get_curl_private(handle);
+                if (cd)
+                        chunk_data_free(cd);
+
+                curl_easy_cleanup(handle);
+        }
+
+        while ((handle = queue_pop(dl->ready))) {
+                cd = get_curl_private(handle);
+                if (cd)
+                        chunk_data_free(cd);
+
+                curl_easy_cleanup(handle);
+        }
+
+        while ((handle = queue_pop(dl->completed))) {
+                cd = get_curl_private(handle);
+                if (cd)
+                        chunk_data_free(cd);
+
+                curl_easy_cleanup(handle);
+        }
+
+        free(dl->store_url);
+        queue_free(dl->ready);
+        queue_free(dl->inprogress);
+        queue_free(dl->completed);
+        curl_multi_cleanup(dl->multi);
+        ca_remote_unref(dl->remote);
+
+        free(dl);
+}
+
+static inline void ca_chunk_downloader_freep(CaChunkDownloader **dl) {
+        ca_chunk_downloader_free(*dl);
+}
+
+static CaChunkDownloader *ca_chunk_downloader_new(CaRemote *rr, const char *store_url) {
+        CaChunkDownloader *dl = NULL;
+        int r, i;
+
+        dl = new0(CaChunkDownloader, 1);
+        if (!dl)
+                goto fail;
+
+        dl->remote = ca_remote_ref(rr);
+
+        r = make_curl_multi_handle(&dl->multi);
+        if (r < 0)
+                goto fail;
+
+        dl->ready = queue_new();
+        if (!dl->ready)
+                goto fail;
+
+        dl->inprogress = queue_new();
+        if (!dl->inprogress)
+                goto fail;
+
+        dl->completed = queue_new();
+        if (!dl->completed)
+                goto fail;
+
+        for (i = 0; i < arg_max_active_chunks; i++) {
+                CURL *handle = NULL;
+                ChunkData *cd = NULL;
+
+                cd = chunk_data_new();
+                if (!cd)
+                        goto fail;
+
+                r = make_curl_easy_handle(&handle, write_chunk, &cd->buffer, cd);
+                if (r < 0)
+                        goto fail;
+
+                queue_push(dl->ready, handle);
+        }
+
+        dl->store_url = strdup(store_url);
+        if (!dl->store_url)
+                goto fail;
+
+        return dl;
+
+fail:
+        ca_chunk_downloader_free(dl);
+        return NULL;
+}
+
+#define AVERAGE(sum, n) (n > 0 ? sum / n : sum)
```

Comment:
hmm, can we make this () and double evaulation safe please?

```c
#define AVERAGE(sum, n) ({ typeof(sum) _sum = sum; typeof(n) _n = n; _n > 0 ? _sum / _n : _sum; })
```

or something like that?